hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCancelAirdropHandler.java

```diff
@@ -126,15 +134,127 @@ public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckExcept
     @Override
     public void handle(@NonNull final HandleContext context) throws HandleException {
-        var tokensConfig = context.configuration().getConfigData(TokensConfig.class);
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenCancelAirdropOrThrow();
+        configValidation(context.configuration(), op);
+
+        final var accountStore = context.storeFactory().writableStore(WritableAccountStore.class);
+        final var airdropStore = context.storeFactory().writableStore(WritableAirdropStore.class);
+
+        final var pendingAirdropIds = op.pendingAirdrops();
+        final var payer = context.payer();
+        final var payerAccount = getIfUsable(payer, accountStore, context.expiryValidator(), INVALID_ACCOUNT_ID);
+        validateTrue(payerAccount.hasHeadPendingAirdropId(), INVALID_TRANSACTION_BODY);
+
+        validatePendingAirdropIds(context, pendingAirdropIds, payer, accountStore, airdropStore);
+        deleteAccountAirdropsFromAccount(pendingAirdropIds, airdropStore, payerAccount, accountStore);
+        updateAccountsNumberOfPendingAirdrops(payerAccount, pendingAirdropIds.size(), accountStore);
+        deletePendingAirdropsFromStore(pendingAirdropIds, airdropStore);
+    }
+
+    private static void configValidation(Configuration configuration, TokenCancelAirdropTransactionBody op) {
+        var tokensConfig = configuration.getConfigData(TokensConfig.class);
         validateTrue(tokensConfig.cancelTokenAirdropEnabled(), NOT_SUPPORTED);
+        validateFalse(
+                op.pendingAirdrops().size() > tokensConfig.maxAllowedPendingAirdropsToCancel(),
+                MAX_PENDING_AIRDROP_ID_EXCEEDED);
+    }
+
+    private static void validatePendingAirdropIds(
+            HandleContext context,
+            List<PendingAirdropId> pendingAirdropIds,
+            AccountID payer,
+            WritableAccountStore accountStore,
+            WritableAirdropStore airdropStore) {
+        final var tokenStore = context.storeFactory().readableStore(ReadableTokenStore.class);
+        final var nftStore = context.storeFactory().readableStore(ReadableNftStore.class);
+
+        for (var pendingAirdropId : pendingAirdropIds) {
+            validateTrue(payer.equals(pendingAirdropId.senderIdOrThrow()), INVALID_ACCOUNT_ID);
+            if (pendingAirdropId.hasFungibleTokenType()) {
+                getIfUsable(pendingAirdropId.fungibleTokenTypeOrThrow(), tokenStore);
+            } else {
+                final var nft = pendingAirdropId.nonFungibleTokenOrThrow();
+                validateTrue(nftStore.get(nft) != null, INVALID_NFT_ID);
+            }
+            getIfUsable(
+                    pendingAirdropId.senderIdOrThrow(), accountStore, context.expiryValidator(), INVALID_ACCOUNT_ID);
+            getIfUsable(
+                    pendingAirdropId.receiverIdOrThrow(), accountStore, context.expiryValidator(), INVALID_ACCOUNT_ID);
```

Comment:
Why should we validate receiver here ?

Reply:
I guess to validate that such account exists. Do you think we can remove this check?

Reply:
yeah I think we don't need to worry about receiver, since all we need to validate is the sender is valid. Since, sender is cancelling his airdrops